utils/hdparm.go

```diff
@@ -207,6 +211,136 @@ func (h *Hdparm) DriveCapabilities(ctx context.Context, logicalName string) ([]*
 	return capabilities, err
 }
 
+// WipeDrive implements DriveWipe by calling Sanitize or Erase as appropriate.
+// Sanitize(CryptoErase) is preferred over Sanitize(BlockErase) which is preferred over Erase(CryptographicErase).
+func (h *Hdparm) WipeDrive(ctx context.Context, logger *logrus.Logger, drive *common.Drive) error { // nolint:gocyclo
+	var (
+		esee     bool
+		eseu     bool
+		sanitize bool
+		bee      bool
+		cse      bool
+	)
+	for _, cap := range drive.Capabilities {
+		switch {
+		case cap.Description == "encryption supports enhanced erase":
+			esee = cap.Enabled
+		case cap.Description == "BLOCK ERASE EXT":
+			bee = cap.Enabled
+		case cap.Description == "CRYPTO SCRAMBLE EXT":
+			cse = cap.Enabled
+		case cap.Description == "SANITIZE feature":
+			sanitize = cap.Enabled
+		case strings.HasPrefix(cap.Description, "erase time:"):
+			eseu = strings.Contains(cap.Description, "enhanced")
+		}
+	}
+
+	l := logger.WithField("drive", drive.LogicalName)
+	if sanitize && cse {
+		// nolint:govet
+		l := l.WithField("method", "sanitize").WithField("action", "sanitize-crypto-scramble")
+		l.Info("wiping")
+		err := h.Sanitize(ctx, drive, CryptoErase)
+		if err == nil {
+			return nil
+		}
+		l.WithError(err).Info("failed")
+	}
+	if sanitize && bee {
+		// nolint:govet
+		l := l.WithField("method", "sanitize").WithField("action", "sanitize-block-erase")
+		l.Info("wiping")
+		err := h.Sanitize(ctx, drive, BlockErase)
+		if err == nil {
+			return nil
+		}
+		l.WithError(err).Info("failed")
+	}
+	if esee && eseu {
+		// nolint:govet
+		l := l.WithField("method", "security-erase-enhanced")
+		l.Info("wiping")
+		err := h.Erase(ctx, drive, CryptographicErase)
+		if err == nil {
+			return nil
+		}
+		l.WithError(err).Info("failed")
+	}
+	return ErrIneffectiveWipe
+}
+
+// Sanitize wipes drive using `ATA Sanitize Device` via hdparm --sanitize
+func (h *Hdparm) Sanitize(ctx context.Context, drive *common.Drive, sanact SanitizeAction) error {
+	var sanType string
+	switch sanact { // nolint:exhaustive
+	case BlockErase:
+		sanType = "block-erase"
+	case CryptoErase:
+		sanType = "crypto-scramble"
+	default:
+		return fmt.Errorf("%w: %v", errSanitizeInvalidAction, sanact)
+	}
+
+	verify, err := ApplyWatermarks(drive)
+	if err != nil {
+		return err
+	}
+
+	h.Executor.SetArgs("--yes-i-know-what-i-am-doing", "--sanitize-"+sanType, drive.LogicalName)
+	_, err = h.Executor.Exec(ctx)
+	if err != nil {
+		return err
+	}
+
+	// now we loop until --sanitize-status reports that sanitization is complete
+	for {
```

Comment:
I suggest this loop be external to the library, so as a principle - the library always performs an action, returns the result and allows the caller to check the status of the action. 
This way we'll avoid situations where its unclear whats going on because the library hasn't returned for a while and a context cancellation doesn't cause it to return.

Reply:
I added the loop because when I was testing this out manually `hdparm --sanitize-status` returns progress info (in a weird order) but when I've tried out the example it actually blocks until idle, so this loop might be able to go away. I'll go try and figure out why/how hdparm --sanitize-status does different behavior and may drop this loop. If we keep the loop in and timeout occurs then `.Exec` will return an error and we return from the function.

I'm trying to keep the WipeDrive interface simple to use and reason about. Callers ask us to wipe a drive, we wipe a drive or return an error. I also want to have it so that if we returned successfully then the caller is sure that the drive was wiped, we guarantee that by wrapping the underlying execs around `verify, err := ApplyWatermarks(drive)` and `verify()`.

So unless we want to rework how `WipeDrive` works its not really something to do in this PR. If we did want to do that it would also be a breaking change.

Reply:
Yeah, I theres going to be cases like this and you can refactor iteratively.
In principle though, lets try to have the library return control to the caller asap and leave any polling to the caller.

Reply:
Originally I was working on a workflow like `func DriveWipers() []DriveWiper` and we'd return each sub-method but scrapped it for now. I'd want some ways to get some metadata for each method like `func IsSecure() bool` and/or `func IsFast() bool` with the idea that the caller has more info and can make its own decisions, having it also need to poll for doneness and verify can make sense then too, but idk I still like the simple interface of WipeDrive either wipes the drive or returns an error, if the user wants concurrency they can do it at the disk level with goroutines.